content/blog/monthly-november-2019.md

```diff
@@ -0,0 +1,154 @@
+
+* In [Case study: TiKV in JD Cloud](https://www.cncf.io/blog/2019/11/26/case-study-tikv-in-jd-cloud/), the CNCF talked to JD Cloud about how they use more than 10 TiKV clusters. The largest hosts 20 billion rows, and some clusters see workloads of 40k (50% read/ 50% write) operations per second with latencies under 10 milliseconds.
+* PingCAP CTO and TiKV founding contributor [@c4pt0r] wrote about how TiKV scales horizonally effectively in [Building a Large-scale Distributed Storage System Based on Raft](https://pingcap.com/blog/building-a-large-scale-distributed-storage-system-based-on-raft/).
+
+## Notable PRs
+
+November was a busy month, here's some highlights:
+
+* [@nrc] submitted a great refactoring in [#5857](https://github.com/tikv/tikv/pull/5857), [#5935](https://github.com/tikv/tikv/pull/5935) and [#5964](https://github.com/tikv/tikv/pull/5964), helping us make TiKV more clean and understanable.
+* [@little-wallace] submitted a PR to optimize the conflict check in `prewrite` in [#5846](https://github.com/tikv/tikv/pull/5846).
+* [@little-wallace] also submitted an latches schheduler optimization in [#6094]https://github.com/tikv/tikv/pull/6094).
+* Numerous vectorization PRs were submitted, enabling even faster coprocessor execution.
+* [@wangwangwar] introduced in-place byte encoding in the codec, helping reduce the number of allocations needed in [#6061](https://github.com/tikv/tikv/pull/6061)
+* [@brson] moved more code into engine traits as part of our ongoing engine abstraction in [#5790](https://github.com/tikv/tikv/pull/5790) and [#5901](https://github.com/tikv/tikv/pull/5901).
+* [@breeswish] taught TiKV about `WriteRef` which helps avoid allocations in `PointGetter` and `ForwardScanner`.
+* [@sticnarf] opened a PR to unify our read pools into a single multi-level thread pool in [#5828](https://github.com/tikv/tikv/pull/5828). Work is still early stage, but this PR is very exciting!
+* [@hoverbear] implemented a suggestion from [@siddontang] to reduce our binary sizes while maintaining our backtraces and debugging info in [#5820](https://github.com/tikv/tikv/pull/5820).
+* [@hunterlxt] taught TiKV to use GRPC's memory quota features in [#5818](https://github.com/tikv/tikv/pull/5818).
+* [@glorv] and [@hoverbear] are working on packaging TiKV.
+
+## Notable issues
+
+> TLA+ was firstly used by PingCAP in 2017, to verify a variant of Percolator transaction model is feasible. Different from original Percolator, TiDB optimizes its efficiency by supporting primary key and secondary keys prewriting concurrently in prewrite phase. It was a problematic idea at first glance, but with TLA+ the case was clearly identified and targeted changes were applied to enable this optimization. Another example is TLA+ verified the correctness of multi-raft region merge algorithm, which gives confidence before implementing and summarized a more clear documentation. These work was made by a 4-member group and proved a worthy investing afterwards.
```

Comment:
Use the below version... The author asked me to edit it days ago but I forgot to reply. 

TLA+ was firstly used by PingCAP in 2017 to verify the feasibility of a variant of the Percolator transaction model. Different from the original Percolator, TiDB’s model has improved efficiency greatly by supporting primary key and secondary key prewriting concurrently in the prewrite phase. The implementation was problematic at first. Thanks to TLA+, the root cause was clearly identified and targeted changes were applied to enable this optimization. Another application of TLA+ at PingCAP is the correctness verification of the multi-raft region merge algorithm, which provided the necessary confidence before we implemented the feature. There were only 4 members invested in the TLA+ related work, but the rewards are far beyond that.

Engineers at PingCAP believe that TLA+ is necessary when it comes to verifying the correctness of distributed systems and specifying the right system behaviors to avoid future patching. They agree that TLA+ is better at expressing the subtle details which natural languages are not good at.

PingCAP has maintained a repo in Github that records all specifications they wrote and keeps aligning TLA+ specifications with the newly implemented database optimizations. Up to now, there're still active PRs and discussions on verifying new features and bugfixes. Speeches and posts on technical reflections and experience of TLA+ application have been shared with the infra community by PingCAP.